### ISSUE TITLE
车道线类别识别

### ISSUE THREAD

**Issue by ycdhqzhiai** [cfzd/Ultra-Fast-Lane-Detection#279]
@cfzd 大佬请教下，我这边添加了一个分支进行车道线类别识别(白黄、虚实等)，按照白黄、虚实两两组合，可以有四个类别，然后加上背景类，总计5类。分支结构：
```python
self.index =  torch.nn.Sequential(
    torch.nn.Conv2d(512, 4, 1, bias=False),
    torch.nn.BatchNorm2d(4),
    torch.nn.ReLU()
)

self.lane_cls = torch.nn.Sequential(
    torch.nn.Linear(900, 1024), 
    torch.nn.ReLU(),
    torch.nn.Linear(1024, 4*5),
)
```
forward
```python
i_fea = self.index(fea).view(-1, 900)
index_cls = self.lane_cls(i_fea).view(-1, 5,4)
```
loss
```
loss_dict = {
    'name': ['cls_loss', 'relation_loss', 'aux_loss', 'relation_dis'],
    'op': [SoftmaxFocalLoss(2), ParsingRelationLoss(), torch.nn.CrossEntropyLoss(), ParsingRelationDis(), torch.nn.CrossEntropyLoss()],
    'weight': [1.0, cfg.sim_loss_w, 1.0, cfg.shp_loss_w, 1.0],
    'data_src': [('cls_out', 'cls_label'), ('cls_out',), ('seg_out', 'seg_label'), ('cls_out',), ('index_out', 'index_label')]
}
```
但是结果好像没起作用，我想请假下大佬，这种一张图片多条车道线，每根车道线多个类别的分类一般怎么处理

**Comment by cfzd** [comment]
@ycdhqzhiai 
我觉得你的处理是没有问题的，可能需要看看给的数据是否正确？或者提供一些最后一个分类loss的变化情况？

如果实在不行的话也可以尝试多标签分类。

**Comment by ycdhqzhiai** [comment]
已解决，loss_dict names少写了类别name，导致没有计算其loss，另外修改了最后shape为B*4*5

**Comment by xiaozhangniaibu** [comment]
> @cfzd 大佬请教下，我这边添加了一个分支进行车道线类别识别(白黄、虚实等)，按照白黄、虚实两两组合，可以有四个类别，然后加上背景类，总计5类。分支结构：
> 
> ```python
> self.index =  torch.nn.Sequential(
>     torch.nn.Conv2d(512, 4, 1, bias=False),
>     torch.nn.BatchNorm2d(4),
>     torch.nn.ReLU()
> )
> 
> self.lane_cls = torch.nn.Sequential(
>     torch.nn.Linear(900, 1024), 
>     torch.nn.ReLU(),
>     torch.nn.Linear(1024, 4*5),
> )
> ```
> 
> forward
> 
> ```python
> i_fea = self.index(fea).view(-1, 900)
> index_cls = self.lane_cls(i_fea).view(-1, 5,4)
> ```
> 
> loss
> 
> ```
> loss_dict = {
>     'name': ['cls_loss', 'relation_loss', 'aux_loss', 'relation_dis'],
>     'op': [SoftmaxFocalLoss(2), ParsingRelationLoss(), torch.nn.CrossEntropyLoss(), ParsingRelationDis(), torch.nn.CrossEntropyLoss()],
>     'weight': [1.0, cfg.sim_loss_w, 1.0, cfg.shp_loss_w, 1.0],
>     'data_src': [('cls_out', 'cls_label'), ('cls_out',), ('seg_out', 'seg_label'), ('cls_out',), ('index_out', 'index_label')]
> }
> ```
> 
> 但是结果好像没起作用，我想请假下大佬，这种一张图片多条车道线，每根车道线多个类别的分类一般怎么处理

老哥neng

> @cfzd 大佬请教下，我这边添加了一个分支进行车道线类别识别（白黄、虚实等），按照白黄、虚实两两组合，可以有四个类别，然后加上背景类，总计5类。分支结构：
> 
> ```python
> self.index =  torch.nn.Sequential(
>     torch.nn.Conv2d(512, 4, 1, bias=False),
>     torch.nn.BatchNorm2d(4),
>     torch.nn.ReLU()
> )
> 
> self.lane_cls = torch.nn.Sequential(
>     torch.nn.Linear(900, 1024), 
>     torch.nn.ReLU(),
>     torch.nn.Linear(1024, 4*5),
> )
> ```
> 
> 向前
> 
> ```python
> i_fea = self.index(fea).view(-1, 900)
> index_cls = self.lane_cls(i_fea).view(-1, 5,4)
> ```
> 
> 损失
> 
> ```
> loss_dict = {
>     'name': ['cls_loss', 'relation_loss', 'aux_loss', 'relation_dis'],
>     'op': [SoftmaxFocalLoss(2), ParsingRelationLoss(), torch.nn.CrossEntropyLoss(), ParsingRelationDis(), torch.nn.CrossEntropyLoss()],
>     'weight': [1.0, cfg.sim_loss_w, 1.0, cfg.shp_loss_w, 1.0],
>     'data_src': [('cls_out', 'cls_label'), ('cls_out',), ('seg_out', 'seg_label'), ('cls_out',), ('index_out', 'index_label')]
> }
> ```
> 
> 但是结果好像没起作用，我想请假下大佬，这种一张图片多条车道线，每根车道线多个类别的分类一般怎么处理

老哥我最近也在研究车道线实虚和白黄的分类，能加个好友联系一下吗

**Comment by ycdhqzhiai** [comment]
可以邮件ycdhqzhiai@163.com
